jib-core/src/main/java/com/google/cloud/tools/jib/registry/ManifestPusher.java

```diff
@@ -53,6 +58,62 @@ public BlobHttpContent getContent() {
     return Collections.emptyList();
   }
 
+  @Override
+  public Void handleHttpResponseException(HttpResponseException httpResponseException)
+      throws HttpResponseException, RegistryErrorException {
+    // docker registry 2.0 and 2.1 returns:
+    //   400 Bad Request
+    //   {"errors":[{"code":"TAG_INVALID","message":"manifest tag did not match URI"}]}
+    // docker registry:2.2 returns:
+    //   400 Bad Request
+    //   {"errors":[{"code":"MANIFEST_INVALID","message":"manifest invalid","detail":{}}]}
+    // quay.io returns:
+    //   415 UNSUPPORTED MEDIA TYPE
+    //   {"errors":[{"code":"MANIFEST_INVALID","detail":
+    //   {"message":"manifest schema version not supported"},"message":"manifest invalid"}]}
+
+    if (httpResponseException.getStatusCode() != HttpStatus.SC_BAD_REQUEST
+        && httpResponseException.getStatusCode() != HttpStatus.SC_UNSUPPORTED_MEDIA_TYPE) {
+      throw httpResponseException;
+    }
+
+    // TODO turn deserializing the error-code into a library method
+
+    // Obtain the error response code.
+    String errorContent = httpResponseException.getContent();
+    if (errorContent == null) {
+      throw httpResponseException;
+    }
+
+    try {
+      ErrorResponseTemplate errorResponse =
+          JsonTemplateMapper.readJson(errorContent, ErrorResponseTemplate.class);
+      List<ErrorEntryTemplate> errors = errorResponse.getErrors();
+      if (errors.size() == 1) {
+        String errorCodeString = errors.get(0).getCode();
+        if (errorCodeString == null) {
+          // Did not get an error code back.
+          throw httpResponseException;
+        }
+        ErrorCodes errorCode = ErrorCodes.valueOf(errorCodeString);
+        if (errorCode.equals(ErrorCodes.MANIFEST_INVALID)
+            || errorCode.equals(ErrorCodes.TAG_INVALID)) {
+          throw new RegistryErrorExceptionBuilder(getActionDescription(), httpResponseException)
+              .addReason("Registry does not support Image Manifest Version 2, Schema 2")
```

Comment:
I wonder if it's better to be a little bit less assertive with the claim, like "Registry may not be supporting ...". Just thinking `TAG_INVALID` might imply something different in some cases or there were genuinely other reasons (e.g., bugs) that produced an invalid manifest. WDYT?

Reply:
Yeah, for example something like this would have been misleading for the artifactory errors, where the issue wasn't that they didn't support v2.2 schema, the issue was that they were expecting fields we didn't provide.

Reply:
Makes sense.  Note that the original Artifactory error report seemed to be due to [Artifactory not being configured to support v2-2 out of the box](
https://github.com/GoogleContainerTools/jib/issues/534#issuecomment-404679808).

It's too bad the servers don't return any actionable information to identify the registry type.  Docker Registry has the following headers:
```
{content-length=[79], content-type=[application/json; charset=utf-8], date=[Wed, 05 Sep 2018 18:12:25 GMT], docker-distribution-api-version=[registry/2.0], connection=[close]}
```
And `quay.io` has:
```
Headers: {content-length=[131], content-type=[application/json], date=[Wed, 05 Sep 2018 18:13:03 GMT], server=[nginx/1.14.0], connection=[keep-alive]}
```